### PR TITLE
feat: Remove the need of explicitly passing a DataModel to read from Postgres

### PR DESCRIPTION
**Pull request by hopeswiller**
## Overview

Support for dynamicio to be able to read input data model schema from the [schema yml files](https://github.com/miodo/dynamicio/blob/3694785473/demo/resources/schemas/input/bar.yaml), rather than explicitly defining the schema class as per Issue-https://github.com/VorTECHsa/dynamicio/issues/6

## Key Changes

* Added _generate_model_from_schema method to `dynamicio/mixins.py`
* Removed BarDataModel class in demo/src/io.py
* Removed model argument when reading postgres

## Impact

* `Added support to read data model from schema yml files when reading postgres--> feature`

## Checklist

- [x] Appropriate unit tests added/updated
- [x] Module, function, class docstrings updated
- [x] Interface documentation updated

### CODE REVIEW

**Review by KitBurgess — APPROVED** [review]
These changes look good from my perspective.

@Christos-Hadjinikolis the code owner has much deeper understanding of the project!

**Comment by miodohelper** [comment]
Closes https://github.com/miodo/dynamicio/issues/2

File: README.md
Comment (Christos-Hadjinikolis) [inline README.md:729]: Great update to the dos.

`nit`: small typo "andtheir" --> "and their"

File: dynamicio/config.py
Comment (Christos-Hadjinikolis) [inline dynamicio/config.py:202]: Nice, I take it this is used to identify the `table_name` correctly. 👍🏼

File: dynamicio/mixins.py
Comment (Christos-Hadjinikolis) [inline dynamicio/mixins.py:43]: Cool

File: dynamicio/mixins.py
Comment (Christos-Hadjinikolis) [inline dynamicio/mixins.py:684]: `if model and sql_query:` --> correctly removed. 

`if model is None and sql_query is None:` --> correctly removed. 

`if self.options.get("model"):` --> we should remove this too. No need to tell users that this is no longer how `dynamicio` works. If one provides this `kwarg` it will be ignored and they will get a warning about it. 

Note that this is not a breaking change. Repositories that use dynamicio  with a model kwarg will still work fine with this enhancement (with the exception of warnings). 

Now, interacting with postgres can obviously be done in two ways: 
1. With a data model (which in this PR we are implicitly generating from the schema);
2. Through the provision of an sql query. 

Default behaviour is 1.  I see we are accounting for 2 conditionally if an `sql_query` is present. So we are good here.

File: dynamicio/mixins.py
Comment (Christos-Hadjinikolis) [inline dynamicio/mixins.py:687]: Nice, accounting for the case where a schema is not explicitly provided as a definition.

File: dynamicio/mixins.py
Comment (Christos-Hadjinikolis) [inline dynamicio/mixins.py:712]: Please include type annotations in function definitions and remove them from docstrings so that we are aligned with how we do it elsewhere in the project. 

```python
def _generate_model_from_schema(schema_dict: dict, schema_name: str) -> DeclarativeMeta:
     ...
```

Also please remove the docstring as this is not part of the public interface.

File: dynamicio/mixins.py
Comment (Christos-Hadjinikolis) [inline dynamicio/mixins.py:722]: Please rename x to either `token` or `word`.

File: dynamicio/mixins.py
Comment (Christos-Hadjinikolis) [inline dynamicio/mixins.py:736]: Left over?

File: dynamicio/mixins.py
Comment (Christos-Hadjinikolis) [inline dynamicio/mixins.py:737]:
```ptyhon
raise ValueError("A model must be provided")
```

I think this can go away too.

File: dynamicio/mixins.py
Comment (Christos-Hadjinikolis) [inline dynamicio/mixins.py:780]: Redundant as per above comment. Even if we keep it in this will never be true after removing  live 719:
```python
@allow_options({"model"})
def _write_to_postgres(self, df: pd.DataFrame):
```

`model` is no longer an allowed option.

File: dynamicio/mixins.py
Comment (Christos-Hadjinikolis) [inline dynamicio/mixins.py:719]: Nicely caught, but this needs to also be updated here: https://github.com/VorTECHsa/dynamicio/blob/master/dynamicio/mixins.py#L700 to 
```python
@allow_options(pd.read_sql)
```

File: tests/test_mixins.py
Comment (Christos-Hadjinikolis) [inline tests/test_mixins.py:1441]: Nice. Thanks

File: tests/test_mixins.py
Comment (Christos-Hadjinikolis) [inline tests/test_mixins.py:1488]: Rename to "test_read_from_postgres_by_implicitly_generating_datamodel_from_schema"

File: tests/test_mixins.py
Comment (Christos-Hadjinikolis) [inline tests/test_mixins.py:1502]: Remove this test please.

File: tests/test_mixins.py
Comment (Christos-Hadjinikolis) [inline tests/test_mixins.py:1562]: Please use a single assertion: 

```python
assert  len(model.__table__.columns) == len(schema) and  model.__tablename__ == schema_name
```

File: tests/test_mixins.py
Comment (Christos-Hadjinikolis) [inline tests/test_mixins.py:1583]: Ok, this one is on me and I am trying to fix it where I can. 

I overuse fixtures at the expense of a test's readability. Instead, I should be creating fixtures only when instantiating an input takes time and it is repeated in many tests. Here, I think this is not the case. 

Again, it is my fault for promoting this bad practice, but kindly, can we avoid using fixtures here and instead replace `model_expected_columns ` with `[PgModel.id, PgModel.foo, PgModel.bar, PgModel.baz]` directly?  I think it improves readability.

File: tests/test_mixins.py
Comment (Christos-Hadjinikolis) [inline tests/test_mixins.py:1597]: Please remove as it is redundant.

File: tests/test_mixins.py
Comment (Christos-Hadjinikolis) [inline tests/test_mixins.py:2006]: Test `test_when_reading_from_postgres_only_the_model_option_is_considered ` can also go away.

**Review by Christos-Hadjinikolis — CHANGES_REQUESTED** [review]
Great work; left some comments.

File: dynamicio/config.py
Comment (hopeswiller) [reply dynamicio/config.py:202]: yeah exactly.

**Comment by hopeswiller** [comment]
@Christos-Hadjinikolis
I just pushed the new requested changes

File: README.md
Comment (hopeswiller) [reply README.md]: yeah sorry my bad

File: tests/test_mixins.py
Comment (hopeswiller) [reply tests/test_mixins.py]: Test removed

File: dynamicio/mixins.py
Comment (hopeswiller) [reply dynamicio/mixins.py]: Great!

File: tests/test_mixins.py
Comment (hopeswiller) [reply tests/test_mixins.py]: okay, correction has been made

File: README.md
Comment (Christos-Hadjinikolis) [inline README.md:729]: Nice.

File: docs/mixins.html
Comment (Christos-Hadjinikolis) [inline docs/mixins.html:64]: Nice.

File: docs/mixins.html
Comment (Christos-Hadjinikolis) [inline docs/mixins.html]: Thanks

File: dynamicio/mixins.py
Comment (Christos-Hadjinikolis) [inline dynamicio/mixins.py:684]: Nice.

File: dynamicio/mixins.py
Comment (Christos-Hadjinikolis) [inline dynamicio/mixins.py]: Thanks.

File: dynamicio/mixins.py
Comment (Christos-Hadjinikolis) [inline dynamicio/mixins.py:710]: Cool

File: dynamicio/mixins.py
Comment (Christos-Hadjinikolis) [inline dynamicio/mixins.py:697]: Nice

File: dynamicio/mixins.py
Comment (Christos-Hadjinikolis) [inline dynamicio/mixins.py]: Cool.

File: tests/conftest.py
Comment (Christos-Hadjinikolis) [inline tests/conftest.py]: Nice.

**Review by Christos-Hadjinikolis — APPROVED** [review]
LGTM

**Comment by Christos-Hadjinikolis** [comment]
@hopeswiller let me merge this once I have tested a release candidate.

File: dynamicio/mixins.py
Comment (Christos-Hadjinikolis) [inline dynamicio/mixins.py:43]: So, it seems like we have an issue here. We are missing a type and that's `DateTime`. 

I think it needs to be added. The error I got from trying the release candidate was not very clear, as it was a key-error. But I think this is it.

**Review by Christos-Hadjinikolis — CHANGES_REQUESTED** [review]
We appear to have an issue with dtypes.

File: dynamicio/mixins.py
Comment (Christos-Hadjinikolis) [reply dynamicio/mixins.py:43]: I am pretty sure this is the problem. I can confirm that I was able to read other tables which did not have that dtype. 

Still to confirm we can write. Let's get this fixed and we can try again.

Comment (Christos-Hadjinikolis) [reply dynamicio/mixins.py:43]: FYI @KitBurgess @hopeswiller

Comment (Christos-Hadjinikolis) [reply dynamicio/mixins.py:43]: Code broke here: https://github.com/miodo/dynamicio/blob/enhancement/implicit_read_model/dynamicio/core.py#L268 

with a key-error for a column that has dateime as the type.

Comment (Christos-Hadjinikolis) [reply dynamicio/mixins.py:43]: Please add a test to confirm we can use all dtypes as welll.

Comment (hopeswiller) [reply dynamicio/mixins.py:43]: In that case I suppose the type specified in the scheme definition yaml file was either timestamptz or date.
If not can you please confirm what the type was

Comment (Christos-Hadjinikolis) [reply dynamicio/mixins.py:43]: Screenshot 2022-07-06 at 15 01 47

Comment (Christos-Hadjinikolis) [reply dynamicio/mixins.py:43]: Assume a yaml schema as follows: 
Screenshot 2022-07-06 at 15 05 09

Which had tobe converted as per below:
Screenshot 2022-07-06 at 15 03 53

**Comment by hopeswiller** [comment]
@Christos-Hadjinikolis  @KitBurgess 
 I have pushed new changes i.e adding extra types to the type lookup dict and test case.
I guess i have to create another release candidate then.

**Comment by Christos-Hadjinikolis** [comment]
> a release candidate is available now.
> 
> On standby for any feedback

Will get back to asap.

**Review by Christos-Hadjinikolis — APPROVED** [review]
Great it works! Read and write. Merging!

File: dynamicio/core.py
Comment (KitBurgess) [inline dynamicio/core.py:270]: I assume these were used for debugging, and we want to remove these print statements?

Comment (hopeswiller) [reply dynamicio/core.py:270]: yeah sorry about that.
I added that when i was debugging the dtypes. I will remove them

Comment (Christos-Hadjinikolis) [reply dynamicio/core.py:270]: Thanks @KitBurgess I missed them and merged had to force-push a fix earlier today. It's all good now guys.

Comment (hopeswiller) [reply dynamicio/core.py:270]: Great! thanks.
Then no need resolve it on my end

**Pull request merged**
